contracts/Pool.sol

```diff
@@ -28,3 +28,83 @@ contract Pool {
+    /// Iterating over the whole list removes all the zeroed items.
+    /// @param current The previously returned receiver address or zero to start iterating
+    /// @return next The next receiver address
+    /// @return weight The next receiver weight, zero if the end of the list was reached
+    function nextWeight(
+        mapping(address => ReceiverWeight) storage self,
+        address current
+    ) internal returns (address next, uint32 weight) {
+        ReceiverWeight storage currentReceiverWeight = self[current];
+        next = currentReceiverWeight.next;
+        weight = 0;
+        if (next != ADDR_END && next != ADDR_UNINITIALIZED) {
+            ReceiverWeight storage nextReceiverWeight = self[next];
+            weight = nextReceiverWeight.weight;
+            // remove elements being zero
+            if (weight == 0) {
+                do {
+                    next = nextReceiverWeight.next;
+                    // delete nextReceiverWeight to release its memory
+                    nextReceiverWeight.next = ADDR_UNINITIALIZED;
+                    if (next == ADDR_END) break;
+                    nextReceiverWeight = self[next];
+                    weight = nextReceiverWeight.weight;
+                } while (weight == 0);
+                // link the previous non-zero element with the next non-zero element
+                // or ADDR_END if it became the last element on the list
+                currentReceiverWeight.next = next;
+            }
+        }
+    }
+
+    /// @notice Get weight for a specific receiver
+    /// @param receiver The receiver to get weight
+    /// @return weight The receinver weight
+    function getWeight(
+        mapping(address => ReceiverWeight) storage self,
+        address receiver
+    ) internal view returns (uint32 weight) {
+        weight = self[receiver].weight;
+    }
+
+    /// @notice Set weight for a specific receiver
+    /// @param receiver The receiver to set weight
+    /// @param weight The weight to set
+    /// @return previousWeight The previously set weight, may be zero
+    function setWeight(
+        mapping(address => ReceiverWeight) storage self,
+        address receiver,
+        uint32 weight
+    ) internal returns (uint32 previousWeight) {
+        ReceiverWeight storage receiverWeight = self[receiver];
+        previousWeight = receiverWeight.weight;
+        receiverWeight.weight = weight;
+        // Item not attached to the list
+        if (receiverWeight.next == ADDR_UNINITIALIZED) {
+            ReceiverWeight storage rootReceiverWeight = self[ADDR_ROOT];
+            address rootNext = rootReceiverWeight.next;
+            rootReceiverWeight.next = receiver;
+            // The first item ever added to the list, root item not initialized yet
+            if (rootNext == ADDR_UNINITIALIZED) rootNext = ADDR_END;
+            receiverWeight.next = rootNext;
+        }
+    }
+}
```

Comment:
So this is basically implemented as a linked list. I worry that this is expensive and complex. Can we not use a flat array?

Reply:
Could you explain? How a flat array would be cheaper to manage?

Reply:
A linked list forces us to keep an extra `next` pointer per item, and a lookup for every iteration. I'm wondering if traversing a flat array would be cheaper and result in simpler code as well :man_shrugging:

Reply:
Iterating is easier, but modifications become way harder. How to update a weight for address `X`? How to tell update from creation of a new item? How to delete an item? Each of these operations would require iteration over the whole or at least an arbitrary part of the list. The mapping-linked list hybrid solves both iteration and modification use cases.